### PR TITLE
Support ReduceLROnPlateau scheduler

### PR DESCRIPTION
**Pull request by jpchen**
This PR:
1) changes the scheduler interface to be more similar to PyTorch's (and changes the order of the calls as per 1.1)
2) supports the remaining schedulers (esp ReduceLROnPlateau [which has been requested](https://forum.pyro.ai/t/properly-using-a-scheduler-with-an-optimizer/730/4))

there are still some outstanding bugs that are being fixed on the pytorch end.

Tested:
- schedulers that have different interfaces
- fixed and variable epochs

### CODE REVIEW

File: pyro/optim/lr_scheduler.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ class PyroLRScheduler(PyroOptim):
         pyro_scheduler = pyro.optim.ExponentialLR({'optimizer': optimizer, 'optim_args': {'lr': 0.01}, 'gamma': 0.1})
         svi = SVI(model, guide, pyro_scheduler, loss=TraceGraph_ELBO())
         svi.step()
+        pyro_scheduler.step(epoch=epoch)
```
Comment (martinjankowiak) [inline pyro/optim/lr_scheduler.py]: can we please update this to be a full and complete snippet with comments, something like
```
for epoch in range(num_epochs):
    blah() # this happens once an epoch
    for minibatch in loader:
        svi.step() # this takes single gradient step and generally happens many times per epoch
```

Comment (martinjankowiak) [reply pyro/optim/lr_scheduler.py]: also please double-check the docstring for correctness/completeness in light of changes

File: pyro/optim/optim.py [hunk 2]
```diff
@@ -51,12 +51,12 @@ def __call__(self, params,  *args, **kwargs):
                     state = self._state_waiting_to_be_consumed.pop(param_name)
                     self.optim_objs[p].load_state_dict(state)
 
-            # actually perform the step for the optim object
-            self.optim_objs[p].step(*args, **kwargs)
-
-            # if optim object was a scheduler, perform an actual optim step
-            if isinstance(self.optim_objs[p], torch.optim.lr_scheduler._LRScheduler):
+            if isinstance(self.optim_objs[p], torch.optim.lr_scheduler._LRScheduler) or \
```
Comment (martinjankowiak) [inline pyro/optim/optim.py]: why is this branch necessary

Comment (jpchen) [reply pyro/optim/optim.py]: we only want to step the optimizer, not the scheduler here. this branch is if `self.optim_objs[p]` is a scheduler.

File: pyro/optim/lr_scheduler.py [hunk 3]
```diff
@@ -5,20 +5,22 @@
...
 
     Example::
 
         optimizer = torch.optim.SGD
-        pyro_scheduler = pyro.optim.ExponentialLR({'optimizer': optimizer, 'optim_args': {'lr': 0.01}, 'gamma': 0.1})
+        scheduler = pyro.optim.ExponentialLR({'optimizer': optimizer, 'optim_args': {'lr': 0.01}, 'gamma': 0.1})
         svi = SVI(model, guide, pyro_scheduler, loss=TraceGraph_ELBO())
-        svi.step()
-        pyro_scheduler.step(epoch=epoch)
+        for i in range(epochs):
```
Comment (martinjankowiak) [inline pyro/optim/lr_scheduler.py]: i => epoch?

File: tests/optim/test_optim.py [hunk 4]
```diff
@@ -97,20 +99,32 @@ def guide():
 
     svi = SVI(model, guide, scheduler, loss=TraceGraph_ELBO())
     for epoch in range(4):
+        svi.step()
         svi.step()
         loc = pyro.param('loc').unconstrained()
         opt_loc = scheduler.optim_objs[loc].optimizer
         opt_scale = scheduler.optim_objs[loc].optimizer
+        if scheduler.pt_scheduler_constructor is torch.optim.lr_scheduler.ReduceLROnPlateau:
```
Comment (fritzo) [inline tests/optim/test_optim.py]: `issubclass`

File: pyro/optim/optim.py (hunk 2 above)
Comment (fritzo) [inline pyro/optim/optim.py]: smells bad to me.

Comment (jpchen) [reply pyro/optim/optim.py]: i'm not sure how to make it cleaner given the design choice. we just want to step the optimizers here.  `optim_objs` is a dict of optimizers _or_ a dict of schedulers each of which is wrapping an optimizer.  in the `lr_scheduler` initialization, we could store the schedulers and optimizers separately, which would simplify this logic but essentially have two references to the optimizer.  WDYT?

Comment (fritzo) [reply pyro/optim/optim.py]: ok, I guess we've inherited this bad smell from PyTorch 😞
